lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2492,18 +2492,20 @@ bool TypeChecker::typeCheckCondition(Expr *&expr, DeclContext *dc) {
   return !resultTy;
 }
 
-bool TypeChecker::typeCheckStmtCondition(StmtCondition &cond, DeclContext *dc,
-                                         Diag<> diagnosticForAlwaysTrue) {
+bool TypeChecker::typeCheckConditionForStatement(LabeledConditionalStmt *stmt,
+                                                 DeclContext *dc) {
   auto &Context = dc->getASTContext();
   bool hadError = false;
   bool hadAnyFalsable = false;
+  auto cond = stmt->getCond();
   for (auto &elt : cond) {
     if (elt.getKind() == StmtConditionElement::CK_Availability) {
       hadAnyFalsable = true;
       continue;
     }
 
     if (auto E = elt.getBooleanOrNull()) {
+      assert(!E->getType() && "the bool condition is already type checked");
```

Comment:
Well, a test case actually hit this. I'm investigating.

Reply:
Added a patch to invalidate `VarDecl`s in patterns if the initializer couldn't type check.

The failing test case was like:
```swift
func test(cond: Bool) {
  guard let val: Int, cond else { ... val ... }
}
```
In this case `cond` is typechecked, but the `VarDecl` for `val` used to remain non-typechecked state.